sound/soc/sof/ops.c

```
@@ -161,6 +161,7 @@ void snd_sof_dsp_panic(struct snd_sof_dev *sdev, u32 offset)
 	sdev->dbg_dump_printed = false;
 
 	snd_sof_dsp_dbg_dump(sdev, SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_MBOX);
+	sof_set_fw_state(sdev, SOF_FW_CRASHED);
```

Consider this firmware boot sequence:
<details>
  <summary>Firmware boot spiced up with a DSP panic</summary>

```
[ 2267.902592] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Turning i915 HDAC power 1
[ 2267.907506] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000044
[ 2267.910443] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000040 at  00000048
[ 2267.910471] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000000 at  00000048
[ 2267.914246] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000048
[ 2267.914273] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000040 at  00000048
[ 2267.915962] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Turning i915 HDAC power 0
[ 2267.916250] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Current DSP power state: D0I0
[ 2267.916256] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state change: 0 -> 1
[ 2267.916262] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: loading firmware
[ 2267.916266] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state change: 1 -> 2
[ 2267.916301] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000002 at  00000048
[ 2267.916327] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000000 at  00000048
[ 2267.916375] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000044
[ 2267.916410] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000004 at  00000044
[ 2267.916415] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: booting DSP firmware
[ 2267.917092] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x40000 successful
[ 2267.917454] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x40000 successful
[ 2267.917488] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: period_bytes:0x0
[ 2267.917494] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: periods:1
[ 2267.917814] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Attempting iteration 0 of Core En/ROM load...
[ 2267.917843] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[ 2267.917892] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f0f successful
[ 2267.917977] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f00 successful
[ 2267.918000] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = f
[ 2267.918058] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask f
[ 2267.918768] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0000 successful
[ 2267.918788] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = 1
[ 2267.918824] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask 1
[ 2267.932773] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0xd4]=0x80000000 successful
[ 2267.932810] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0e0e successful
[ 2267.932829] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x1010e0e successful
[ 2267.932838] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask e
[ 2267.932969] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error : DSP panic!
[ 2267.932977] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: panic: dsp_oops_offset 788480 offset 788480
[ 2267.932985] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ------------[ DSP dump start ]------------
[ 2267.932989] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state: SOF_FW_BOOT_IN_PROGRESS (2)
[ 2267.933007] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: status: fw entered - code 00000005
[ 2267.933206] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: reason: assertion failed (0xb)
[ 2267.933211] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: trace point: 0x000041c0
[ 2267.933215] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: panic at src/schedule/edf_schedule.c:98
[ 2267.933221] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error: DSP Firmware Oops
[ 2267.933225] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EXCCAUSE 0x0000003f EXCVADDR 0x00000000 PS       0x00060025 SAR     0x00000000
[ 2267.933233] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPC1     0x00000000 EPC2     0xbe02be0d EPC3     0x00000000 EPC4    0x00000000
[ 2267.933241] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPC5     0x00000000 EPC6     0x00000000 EPC7     0x00000000 DEPC    0x00000000
[ 2267.933246] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPS2     0x00060c20 EPS3     0x00000000 EPS4     0x00000000 EPS5    0x00000000
[ 2267.933252] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPS6     0x00000000 EPS7     0x00000000 INTENABL 0x00000000 INTERRU 0x00000222
[ 2267.933284] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: stack dump from 0xbe086140
[ 2267.933291] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086140: 00000000 00000000 00000000 00000000
[ 2267.933298] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086150: 00000000 00000000 00000000 00000000
[ 2267.933319] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086160: be086140 00000000 00000000 00000000
[ 2267.933328] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086170: 00000000 00000000 00000000 00000000
[ 2267.933335] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086180: 00000000 00000000 00000000 00000000
[ 2267.933342] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086190: 00000000 00000000 2f637273 65686373
[ 2267.933347] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe0861a0: 656c7564 6664652f 6863735f 6c756465
[ 2267.933352] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe0861b0: 00632e65 00000000 00000062 ffffffff
[ 2267.933356] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ------------[ DSP dump end ]------------
[ 2268.233472] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5 timedout
[ 2268.233507] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x1010f0f successful
[ 2268.233713] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f successful
[ 2268.233721] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[ 2268.233726] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Attempting iteration 1 of Core En/ROM load...
[ 2268.233744] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[ 2268.234405] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f0f successful
[ 2268.234434] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f00 successful
[ 2268.234444] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = f
[ 2268.234456] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask f
[ 2268.234479] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0000 successful
[ 2268.234487] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = 1
[ 2268.234498] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask 1
[ 2268.235377] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0xd4]=0x80000000 successful
[ 2268.235405] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0e0e successful
[ 2268.235702] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x1010e0e successful
[ 2268.235709] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask e
[ 2268.324348] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5000001 successful
[ 2268.341431] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5 successful
[ 2268.341460] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x140000 successful
[ 2268.341476] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Firmware download successful, booting...
[ 2268.363607] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx: 0x70000000: FW_READY
[ 2268.363620] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ipc: DSP is ready 0x70000000 offset 0x81000
[ 2268.363627] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state change: 2 -> 5
[ 2268.363643] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx done: 0x70000000: FW_READY
[ 2268.363708] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: firmware boot complete
[ 2268.363739] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000048
[ 2268.363760] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000002 at  00000048
[ 2268.363773] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000004 at  00000044
[ 2268.363782] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000000 at  00000044
[ 2268.363809] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x40000 successful
[ 2268.363847] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x40000 successful
[ 2268.363857] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: period_bytes:0x0
[ 2268.363862] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: periods:1
[ 2268.363891] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: snd_sof_init_trace_ipc: stream_tag: 1
```
I have already re-considered to mark the firmware crashed when an IPC timeout happens as it is part of a 'normal' operation and just an indication of a bad alignment of events.
Now it looks like that when we receive a **DSP panic!**, it is not that fatal as it sounds, it is just again part of the ways the firmware operates.
Iow, when the firmware screams that I have crashed, it is not necessarily true.

With this level of consistency we could as well give up on handling firmware errors on the kernel side as there is no way to tell if it is crying wolf or real.

Gash.

PS: this particular 'panic' originates from this line: https://github.com/thesofproject/sof/blob/1c1243b8900905138021649cfde1e7559f3aa2af/src/schedule/edf_schedule.c#L98

Should we do an audit of `assert()` use and remove the ones that are not errors?

@plbossart, @lgirdwood, @kv2019i, @lyakh, @ranj063, any idea, hint?

---

maybe we throw the error message too early, we have multiple iterations for the boot phase and if we have an error, be it a timeout or a panic, before the end of all the iterations we should probably not let clients know.

This doesn't remove the need to audit wth the firmware uses assert(), which IMHO should be avoided.

---

If I understand this right, this is something which happens regularly and this is why we have the `boot_iteration` counter in hda-loader.c. The DSP just fails to boot without real reason.
With my changes now we are printing the full trace dump in response to a panic event from the firmware, before we only printed the extended rom status only.

We could try to suppress the dumps if we are in `SOF_FW_BOOT_IN_PROGRESS` _and_ the `boot_iteration < HDA_FW_BOOT_ATTEMPTS` _and_ `SOF_DBG_PRINT_ALL_DUMPS` is not set.
Lots of new if, but can be done if needed.

---

what we did before was to keep the logs but demote them to dev_dbg() during the initial iteration. We could use a dev_warn(), but at least not flag the logs as errors.

---

The 'extended rom status' was printed with dev_err() and it was printed if the fw_status was not SOF_FW_BOOT_COMPLETE.
This is different path, this is in response to a DSP panic interrupt and event, not in the boot loop.

---

After a short sleep I think we should retain and print all dumps when a DSP panic is received. It is pointing to a SW bug (assert() / panic() is invoked).
If it happens during firmware boot, that's a bit tricky as we will try the boot again and it might succeed the next time, but it is still a SW asserted panic.

Runtime selecting ERR or WARN level is kind of nasty as the main part of the dump is coming from `snd_sof_get_status()` with wired dev_err() and no flags coming it atm.

So, I'm not sure what should be done, but before my debug changes we were printing with dev_err() also, so this is not new.

---

so the conclusion is what then?

---

@plbossart, I'm not too thrilled to add a new parameter to `snd_sof_dsp_panic()` saying that this panic is not necessarily means that the firmware crashed. Well, it means that the firmware crashed, but we are going to give it another go, so please, don't set the state to crashed yet.
If we don't want to see these by default then we also need another flag to make the retry period prints optional.

I don't have better idea to be honest

---

The problem is that sof-test and CI does not look at the nuances, if there is a dev_err() statement that's reported as a FAIL, even if the boot succeeds eventually. So either we fix sof-test or we avoid a dev_err() for a non-critical issue.

---

Right, we can do that without leaving the hda-ipc.c, we can suppress the dumps until the last retry if the print all dumps is not specified.
But we need the parameter to not mark the state to crashed, so it is better to add the two new parameter.

---

The revised and improved SOF_FW_CRASHED state handling is now merged (#3225) and an update to improve the state handling is waiting (#3231).